src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs

```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
             allArgs.AddRange(references.Select(r => $"-r:\"{r}\""));
             allArgs.AddRange(resources.Select(resource => $"-resource:{resource}"));
             allArgs.AddRange(sourceFiles.Select(s => $"\"{s}\""));
-            allArgs.Prepend($"-noconfig");
+
+            // Gather all compile IO
+            var inputs = new List<string>();
+            var outputs = new List<string>();
+
+            inputs.Add(context.ProjectFile.ProjectFilePath);
+ 
+            if (context.LockFile != null)
+            {
+                inputs.Add(context.LockFile.LockFilePath);
+            }
+
+            if (context.LockFile.ExportFile != null)
+            {
+                inputs.Add(context.LockFile.ExportFile.ExportFilePath);
+            }
+
+            inputs.AddRange(sourceFiles);
+            inputs.AddRange(references);
+            outputs.AddRange(outputPaths.CompilationFiles.All());
```

Comment:
Yes, All() is a public method of the Microsoft.DotNet.ProjectModel.CompilationOutputFiles class, then you get an IEnumerable<string> where you have the .dll output path and also e.g the .pdb output path.